### ISSUE TITLE
drush error with PHP 5.6

### ISSUE THREAD

**Issue by Clemchan33** [drud/ddev#5610]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Output of `ddev debug test`

<details><summary>Expand `ddev debug test` diagnostic information</summary>

```
Running bash [-c /tmp/test_ddev.sh]
======= Existing project config =========
These config files were loaded for project test: [/home/clemchan/Docker/test/.ddev/config.yaml]
name: test
type: drupal7
php_version: 5.6
webserver_type: apache-fpm
webimage: ddev/ddev-webserver:v1.22.5
additional_hostnames: []
additional_fqdns: []
database: {mysql 5.6}
hooks: map[post-import-db:[map[exec:mysql -uroot -proot -e "UPDATE variable SET scytale_url = 'https://test.ddev.site';" service:db] map[exec:mysql -uroot -proot -e "UPDATE variable SET frontoffice_base_url = 'https://test.ddev.site';" service:db]] post-start:[map[exec:composer global require drush/drush:8.x] map[exec:echo 'export PATH="${HOME}/.composer/vendor/bin:$PATH"' >> ~/.bashrc] map[exec:source ~/.bashrc]]]
project_tld: ddev.site
use_dns_when_possible: true
composer_version: 2.2
nodejs_version: 18
default_container_timeout: 120
======= Creating dummy project named  tryddevproject-9135 in ../tryddevproject-9135 =========
OS Information: Linux FoxFly 5.10.16.3-microsoft-standard-WSL2 #1 SMP Fri Apr 2 22:23:49 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
User information: uid=1000(clemchan) gid=1000(clemchan) groups=1000(clemchan),4(adm),20(dialout),24(cdrom),25(floppy),27(sudo),29(audio),30(dip),44(video),46(plugdev),116(netdev),999(docker)
DDEV version:  ITEM             VALUE
 DDEV version     v1.22.5
 architecture     amd64
 db               ddev/ddev-dbserver-mariadb-10.4:v1.22.5
 ddev-ssh-agent   ddev/ddev-ssh-agent:v1.22.5
 docker           24.0.7
 docker-compose   v2.23.3
 docker-platform  wsl2-docker-ce
 mutagen          0.17.2
 os               linux
 router           ddev/ddev-traefik-router:v1.22.5
 web              ddev/ddev-webserver:v1.22.5
PROXY settings: HTTP_PROXY='' HTTPS_PROXY='' http_proxy='' NO_PROXY=''
======= DDEV global info =========
Global configuration:
instrumentation-opt-in=true
omit-containers=[]
performance-mode=none
router-bind-all-interfaces=false
internet-detection-timeout-ms=3000
disable-http2=false
use-letsencrypt=false
letsencrypt-email=
table-style=default
simple-formatting=false
use-hardened-images=false
fail-on-hook-fail=false
required-docker-compose-version=v2.23.3
use-docker-compose-from-path=false
project-tld=
xdebug-ide-location=
no-bind-mounts=false
router=traefik
wsl2-no-windows-hosts-mgt=false
router-http-port=80
router-https-port=443
mailpit-http-port=8025
mailpit-https-port=8026
traefik-monitor-port=10999

======= DOCKER info =========
docker location: -rwxr-xr-x 1 root root 35948624 Oct 26 11:08 /usr/bin/docker
docker version:
Client: Docker Engine - Community
 Version:           24.0.7
 API version:       1.43
 Go version:        go1.20.10
 Git commit:        afdd53b
 Built:             Thu Oct 26 09:08:01 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          24.0.7
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.10
  Git commit:       311b9ff
  Built:            Thu Oct 26 09:08:01 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.25
  GitCommit:        d8f198a4ed8892c764191ef7b3b06d8a2eeb5c7f
 runc:
  Version:          1.1.10
  GitCommit:        v1.1.10-0-g18a0cb0
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
DOCKER_DEFAULT_PLATFORM=notset
======= Mutagen Info =========
======= Docker Info =========
Docker platform: wsl2-docker-ce
Using Docker context: default (unix:///var/run/docker.sock)
docker-compose: v2.23.3
Using DOCKER_HOST=unix:///var/run/docker.sock
Docker version: 24.0.7
Able to run simple container that mounts a volume.
Able to use internet inside container.
Docker disk space:
Filesystem                Size      Used Available Use% Mounted on
overlay                 251.0G     27.6G    210.6G  12% /

 Container ddev-test-db  Stopped
 Container ddev-test-web  Stopped
 Container ddev-test-web  Stopped
 Container ddev-test-db  Stopped
 Container ddev-test-db  Removed
 Container ddev-test-web  Removed
 Network ddev-test_default  Removed
Project test has been stopped.
The ddev-ssh-agent container has been removed. When you start it again you will have to use 'ddev auth ssh' to provide key authentication again.
Network ddev_default removed
Existing docker containers:
CONTAINER ID   IMAGE     COMMAND   CREATED   STATUS    PORTS     NAMES
Network ddev_default created
Starting tryddevproject-9135...
Network ddev-tryddevproject-9135_default created
 Container ddev-ssh-agent  Created
 Container ddev-ssh-agent  Started
ssh-agent container is running: If you want to add authentication to the ssh-agent container, run 'ddev auth ssh' to enable your keys.
v1.22.5: Pulling from ddev/ddev-dbserver-mariadb-10.4
7a2c55901189: Pull complete
b090629ef657: Pull complete
28dc987444b0: Pull complete
42b78737b77d: Pull complete
9859b5389429: Pull complete
aec9a7ae10e5: Pull complete
682574f77cd8: Pull complete
86d3667f0945: Pull complete
421907d821c6: Pull complete
4f4fb700ef54: Pull complete
58c43ccf1498: Pull complete
1c7354ae7b82: Pull complete
6b47deb38700: Pull complete
a41e341cb9d6: Pull complete
08a3c563fa29: Pull complete
237a76340453: Pull complete
a86226c1a22a: Pull complete
c37b25cc6489: Pull complete
5c82817ac12c: Pull complete
f02ed6a4a9bd: Pull complete
3b7a7339b303: Pull complete
c55eee51c293: Pull complete
54be1ba94ea3: Pull complete
f9c09779d6d1: Pull complete
c6805a7e6d0f: Pull complete
Digest: sha256:1d4f38e147898fdc9416562c947b517a5f7030268d0a97ba6476e4f9d416ed3c
Status: Downloaded newer image for ddev/ddev-dbserver-mariadb-10.4:v1.22.5
docker.io/ddev/ddev-dbserver-mariadb-10.4:v1.22.5
Building project images...
Project images built in 2s.
 Container ddev-tryddevproject-9135-web  Created
 Container ddev-tryddevproject-9135-db  Created
 Container ddev-tryddevproject-9135-web  Started
 Container ddev-tryddevproject-9135-db  Started
Waiting for web/db containers to become ready: [web db]
Starting ddev-router if necessary...
 Container ddev-router  Created
 Container ddev-router  Started
Waiting for additional project containers to become ready...
All project containers are now ready.
Successfully started tryddevproject-9135
Project can be reached at https://tryddevproject-9135.ddev.site https://127.0.0.1:32781
======== Curl of site from inside container:
HTTP/1.1 200 OK
Server: nginx
Date: Fri, 08 Dec 2023 09:01:07 GMT
Content-Type: text/html; charset=UTF-8
Connection: keep-alive
Vary: Accept-Encoding

======== curl -I of http://tryddevproject-9135.ddev.site from outside:
HTTP/1.1 200 OK
Content-Type: text/html; charset=UTF-8
Date: Fri, 08 Dec 2023 09:01:07 GMT
Server: nginx
Vary: Accept-Encoding

======== full curl of http://tryddevproject-9135.ddev.site from outside:
Success accessing database... db via TCP/IP
ddev is working. You will want to delete this project with 'ddev delete -Oy tryddevproject-9135'
======== Project ownership on host:
drwxr-xr-x 4 clemchan clemchan 4096 Dec  8 10:00 ../tryddevproject-9135
======== Project ownership in container:
drwxr-xr-x 4 clemchan clemchan 4096 Dec  8 09:00 /var/www/html
======== In-container filesystem:
Filesystem     Type 1K-blocks     Used Available Use% Mounted on
/dev/sdb       ext4 263174212 29775492 219960564  12% /var/www/html
======== curl again of tryddevproject-9135 from host:
Success accessing database... db via TCP/IP
ddev is working. You will want to delete this project with 'ddev delete -Oy tryddevproject-9135'
Thanks for running the diagnostic. It was successful.
Please provide the output of this script in a new gist at gist.github.com
Running ddev launch in 5 seconds
/usr/bin/xdg-open: 882: www-browser: not found
/usr/bin/xdg-open: 882: links2: not found
/usr/bin/xdg-open: 882: elinks: not found
/usr/bin/xdg-open: 882: links: not found
/usr/bin/xdg-open: 882: lynx: not found
/usr/bin/xdg-open: 882: w3m: not found
xdg-open: no method available for opening 'https://tryddevproject-9135.ddev.site'
Failed to run launch ; error=exit status 3

Please run cleanup after debugging with 'ddev debug testcleanup'
Failed running test_ddev.sh: exit status 1
. You can run it manually with `curl -sL -O https://raw.githubusercontent.com/ddev/ddev/master/cmd/ddev/cmd/scripts/test_ddev.sh && bash test_ddev.sh`
```
</details>         

### Expected Behavior

Hi,

On my ddev with PHP 5.6 for Drupal 7, I expect the ddev drush to work.
And also I expect the hooks: post-import-db to wait the importation is finished.

### Actual Behavior

Hi,

On my ddev with PHP 5.6 for Drupal 7, I have got an error with drush :
```
PHP Warning:  Unsupported declare 'strict_types' in phar:///usr/local/bin/drush8/.box/src/Checker.php on line 3

Warning: Unsupported declare 'strict_types' in phar:///usr/local/bin/drush8/.box/src/Checker.php on line 3
PHP Parse error:  syntax error, unexpected ':', expecting ';' or '{' in phar:///usr/local/bin/drush8/.box/src/Checker.php on line 12

Parse error: syntax error, unexpected ':', expecting ';' or '{' in phar:///usr/local/bin/drush8/.box/src/Checker.php on line 12
```

I have made a correction to make drush working inside the web container with : 
```
hooks:
  post-start:
    - exec: composer global require drush/drush:8.x
    - exec: echo 'export PATH="${HOME}/.composer/vendor/bin:$PATH"' >> ~/.bashrc
    - exec: source ~/.bashrc
```

But that doesn't repair the command "ddev drush". It allow only it to work if you go in the web container (with ddev ssh for example).

### Steps To Reproduce

**1. I'm using :** 

- WSL 2 : Ubuntu 22.04.3 LTS
- DDEV 1.22.5
- Docker version 24.0.7, build afdd53b

**2. The config.yaml contains :** 

```
name: test
type: drupal7
docroot: ""
php_version: "5.6"
webserver_type: apache-fpm
xdebug_enabled: false
additional_hostnames: []
additional_fqdns: []
database:
    type: mysql
    version: "5.6"
use_dns_when_possible: true
composer_version: "2.2"
web_environment: []
hooks:
  post-start:
    - exec: composer global require drush/drush:8.x
    - exec: echo 'export PATH="${HOME}/.composer/vendor/bin:$PATH"' >> ~/.bashrc
    - exec: source ~/.bashrc
  post-import-db:
    - exec: mysql -uroot -proot -e "UPDATE variable SET xxxxxxx = 'https://test.ddev.site';"
      service: db
    - exec: mysql -uroot -proot -e "UPDATE variable SET xxxxxxxxxx = 'https://test.ddev.site';"
      service: db
```

(you have to replace xxxxxx with a column of your variable table... sorry I can't give you the correct name)

**3. You have to do :** 

- ddev start
- ddev drush

**Comment by rfay** [comment]
If your project type is drupal7, drush8 is already globally installed.

But I don't think drush8 supports PHP 5.6 any more. You may be able to download and install something really old for drush or use PHP 7.

**Comment by Clemchan33** [comment]
Hi,

Thank you @rfay for your answer but like always, we can't get what we want....
I would prefer be on Drupal 10 with PHP 8.3, but I have to work on Drupal 7 with PHP 5.6.... Not my choice....

How can I change the default drush version used with the command "ddev drush" for my project?

**Comment by rfay** [comment]
You have two chores. 

1. Figure out what Drush version might work with PHP 5.6. You might ask in #drush in Drupal slack. Or you can just try a variety. I'm not sure that old releases like that are available at current location of Drush though. Maybe they're on the old drush project on drupal.org. 
2. Install the one you want. in the web container. Currently in the web container on a drupal7 project, /usr/local/bin/drush is a symlink to /usr/local/bin/drush8, which is current version of drush8. You can use a .ddev/web-build/Dockerfile to rearrange that any way you want. But it won't be a composer install, since all this predates composer.

I would try just downloading a variety of versions of drush into your project, and try using them. Just put them in the root of your project and `ddev ssh` and try using them with an explicit path. For example, download into project root, then `ddev ssh`  and run drush with ./drush to see how it works. 

Details on Dockerfile (once you get a version of drush that works) are at https://ddev.readthedocs.io/en/latest/users/extend/customizing-images/#adding-extra-dockerfiles-for-webimage-and-dbimage - yes, you can also do a post-start hook, it's less efficient but you can do whatever you want to get drush in there.

Good luck!

**Comment by Clemchan33** [comment]
Hi,

Thank you again for your answer.

Well, what I want to know is how can I repair the command `ddev drush`.
What I have done allow me to use drush when I'm in the web container with `ddev ssh` but I can't use it with `ddev drush`.

I will explain a little more, I want to execute command drush after the `ddev import-db`. That's why you see I use a `hook post-import-db`. But with `ddev drush` not working, I can't use drush in my `hook post-import-db`.

**Comment by rfay** [comment]
All `ddev drush` does is run `drush` in the PATH. Nothing else. I guess if you've already found a working version of `drush` you could:
1. Change the link /usr/local/bin/drush to point to your drush instead of to /usr/local/bin/drush8
2. Change the implementation of the drush command in your project. Copy it from ~/.ddev/commands/web/drush to your project .ddev/commands/web/drush and fiddle with it. But I think you'll find there's nothing wrong with it.

This is all it does. It's using drush from the $PATH. It's the same as `ddev exec drush`

```bash
#!/bin/bash

#ddev-generated
## Description: Run drush CLI inside the web container
## Usage: drush [flags] [args]
## Example: "ddev drush uli" or "ddev drush sql-cli" or "ddev drush --version"
## ProjectTypes: drupal7,drupal8,drupal9,drupal10,backdrop
## ExecRaw: true

if ! command -v drush >/dev/null; then
  echo "drush is not available. You may need to 'ddev composer require drush/drush'"
  exit 1
fi
drush "$@"
```

So maybe you just want to change it to use an explicit path for drush, for example, `/var/www/html/some/path/drush "$@"` instead of just `drush "$@"`. Or maybe you want to change the symlink in /usr/local/bin/drush to point to yours. Or maybe you want to add to the in-container PATH to put your path first. All of those will do this.

**Comment by jameswilson** [comment]
# Working solution for old drush on php <=5.6

[OLD Drush docs here](https://docs.drush.org/en/7.x/install/) describe the recommended manual installation procedure using git. This is what I decided to use and follow, along with indications in previous comments on this issue from @rfay.

## Part 1, test manually

1.  Fix filesystem permissions on /usr/local/src inside DDEV web container.

    When you're inside the DDEV web image, the standard installation location cannot be used without `sudo`. Therefore it is recommended to simply change the permissions on this folder before doing anything else.

    ```
    me@host-os$ ddev ssh
    me@ddev-web$ sudo chmod 777 /usr/local/src
    ```

2.  Clone the drush repo and checkout the desired version.

    ```
    me@ddev-web$ git clone https://github.com/drush-ops/drush.git /usr/local/src/drush
    me@ddev-web$ cd /usr/local/src/drush && git checkout 6.7.0
    ```

3.  Replace the drush symlink with the version from source.

    ```
    me@ddev-web$ ls -al /usr/local/bin/drush
    me@ddev-web$ ln -sf /usr/local/src/drush/drush /usr/local/bin/drush
    ```

4.  Install the RobLoach/Console_Table dependency to /usr/local/bin/drush/lib

    Old versions of Drush have additional manual installation steps to follow. You can see this clearly if you forget to run them, drush will tell you.

    ```
    me@ddev-web$ drush --version
    
    Drush needs to download a library from                               [error]
    https://github.com/RobLoach/Console_Table/archive/1.1.5.tar.gz in
    order to function, and the attempt to download this file
    automatically failed because you do not have permission to write to
    the library directory /usr/local/src/drush/lib. To continue you will
    need to manually download the package from
    https://github.com/RobLoach/Console_Table/archive/1.1.5.tar.gz,
    extract it, and copy the directory into your /usr/local/src/drush/lib
    directory.
    
    me@ddev-web$ wget https://github.com/RobLoach/Console_Table/archive/1.1.5.tar.gz -P /usr/local/src/drush/lib
    me@ddev-web$ tar zxvf /usr/local/src/drush/lib/1.1.5.tar.gz -C /usr/local/src/drush/lib/
    ```

6. Confirm Drush works for your version of PHP

    ```
    me@ddev-web:~$ php --version
    PHP 5.6.40-68+0~20230902.80+debian11~1.gbpa2012b (cli)
    Copyright (c) 1997-2016 The PHP Group
    Zend Engine v2.6.0, Copyright (c) 1998-2016 Zend Technologies
        with Zend OPcache v7.0.6-dev, Copyright (c) 1999-2016, by Zend Technologies
    
    me@ddev-web:~$ drush --version
     Drush Version   :  6.7.0
    ```

## Part 2, update Ddev configurations

Officialize the manual testing done in part 1 above into your project's DDEV configuration files. We recommend committing the .ddev folder to standardize DDEV startup with all team members. You may add an abbreviated version of the steps above to the .ddev/web-build/Dockerfile and to the `post-start` hook in .ddev/config.yaml:

Create/update .ddev/web-build/Dockerfile:
```Dockerfile
# Copied from .ddev/web-build/Dockerfile.example
# @see https://github.com/ddev/ddev/issues/5610
ARG BASE_IMAGE
FROM $BASE_IMAGE

# Install old Drush, because version 8 does not work on PHP 5.6
# @see https://docs.drush.org/en/7.x/install/
RUN git clone https://github.com/drush-ops/drush.git /usr/local/src/drush
RUN cd /usr/local/src/drush && git checkout 6.7.0

# Install old Drush dependencies.
RUN wget https://github.com/RobLoach/Console_Table/archive/1.1.5.tar.gz -P /usr/local/src/drush/lib
RUN tar zxvf /usr/local/src/drush/lib/1.1.5.tar.gz -C /usr/local/src/drush/lib/

# Create /usr/local/bin/drush symlink.
# !!!Note!!! this line has no effect as it will be overwritten by Ddev in a later process.
# It must instead be added to .ddev/config.yaml:hooks.post-start
RUN ln -sf /usr/local/src/drush/drush /usr/local/bin/drush
```

Update .ddev/config.yaml:
```yaml
# ...
hooks:
  post-start:
    # @see https://github.com/ddev/ddev/issues/5610
    - exec: ln -sf /usr/local/src/drush/drush /usr/local/bin/drush
# ...
```

## Part 3, some notes

While working through this, I found out a few things along the way:

PHP 5.6 is installed by DDEV via `php_version: "5.6"` in your .ddev/config.yaml. Drush 8 is installed by DDEV via `type: drupal7` in your .ddev/config.yaml.

But when the Dockerfile is executed on `ddev start`, the default PHP version from the Docker image is (currently) 8.1.27 and php 5.6 has not been installed yet. Similarly, drush8 is not  installed and there is nothing at /usr/local/bin/drush. So if you try to add the symlink in the Dockerfile as suggested in step 3 of manual testing above, it will get overridden later by another DDEV processes and set back to point at drush8.

Older Drush versions are likely not compatible with php 8, therefore you'll likely need to do more work to actually **use** OLD `drush` commands from within the Dockerfile once you've installed it. If you need to play with php 5.6 and old drush commands from within Dockerfile, please update this issue with your findings.

tl;dr - The PHP version 5.6 and drush8 are both installed at a later point by DDEV after Dockerfile has executed but before the `post-start` hook in .ddev/config.yaml has executed. So therefore you may use the `post-start` hook in .ddev/config.yaml to override the symlink  which is a relatively inexpensive operation to run on every startup. Use the `-f` option to force override the symlink, eg `ln -sf`.

**Comment by rfay** [comment]
Awesome work, thanks for the docs @jameswilson ! 

A minor note, in current ddev this isn't required in the Dockerfile, and it might be better to call it Dockerfile.old-drush

```
# Copied from .ddev/web-build/Dockerfile.example
# @see https://github.com/ddev/ddev/issues/5610#issuecomment-1946245561
ARG BASE_IMAGE
FROM $BASE_IMAGE
```

If people use this much or it gets activity here, it may be worth considering an add-on, but I'm betting this is fine. It also could go into ddev/ddev-contrib.

**Comment by jameswilson** [comment]
Thanks Randy! Regarding the Dockerfile note, I updated the comment to pull that out into a sidenote:
> Sidenote: Older versions of DDEV also require the following lines at the top of .ddev/web-build/Dockerfile:
> ```Dockerfile
> ARG BASE_IMAGE
> FROM $BASE_IMAGE
> ```

**Comment by cscomfort** [comment]
I recently ran into this same issue, however, please note that Drush 8 _is_ compatible with PHP 5.6 + Drupal 7. See:

https://www.drush.org/12.x/install/#drupal-compatibility

The issue seems to stem from the the version of Drush 8 provided by DDEV. As a quick test, I downloaded the `drush.phar` executable from https://github.com/drush-ops/drush/releases/tag/8.4.9 (chose this version because later ones throw a `stat failed` warning). Next, after making it available in my container as an executable, I confirmed that `drush` commands work as expected. None of the warnings or parsing errors noted as part of this issue were returned.

I'm using this as a workaround for the moment because I won't need to use this DDEV instance for much longer, but wanted to provide some additional context for the folks watching this issue.

**Comment by rfay** [comment]
Current DDEV installs latest Drush 8, 8.4.12. (I think when it moved to PHP 8 it lost PHP 5.6?)

You'll probably see some other problems with 8.4.9 (it had some problems with box IIRC), but of course everybody should use what works for them.

**Comment by cscomfort** [comment]
I just wanted to point out that Drush officially states compatibility with Drush 8 and PHP 5.6. And FWIW, I used 8.4.12 in the steps I outlined above and it did work, there were just some `stat failed` warnings.

All of this is to say that the error reported by the OP is not due to an incompatibility between Drush 8 and PHP 5.6.
